### PR TITLE
benchmark: update HSTU benchmark fixes and docs

### PR DESCRIPTION
**Pull request by JacoCheung**
Summary:
- Fix HSTU FLOP/MFU accounting: dim_per_head in output projection, H100 BF16 peak = 989 TFLOPS, DP-group FLOP reduction.
- Speed up shuffler load balancing with a C++ Karmarkar-Karp partitioner that releases the GIL.
- Reduce Python hot-path stalls with pinned host copies in DynamicEmb and batch shuffler paths.
- Fix benchmark data/timing: one-pass random dataset generation, log-overhead isolation, log/eval/sequence-length knobs, and shorter profiling windows.
- Harden distributed runs: CUDA_DEVICE_MAX_CONNECTIONS default 8, per-rank tmp/cache dirs, DP collective warmup, and configurable GPU request flags.
- Extend benchmark tooling: unified e2e/layer/kernel launch path, layer analyzers, and refreshed benchmark docs.

Validation:
- Benchmark runs completed.
- Pre-commit passed.

### CODE REVIEW

File: corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py [hunk 1]
```diff
@@ -45,6 +45,32 @@
...
+def _copy_cuda_tensor_to_pinned_cpu(tensor: torch.Tensor) -> torch.Tensor:
+    if not tensor.is_cuda:
+        return tensor.detach().cpu()
+    cpu = torch.empty(
+        tuple(tensor.shape),
+        dtype=tensor.dtype,
+        device=torch.device("cpu"),
+        pin_memory=True,
+    )
+    cpu.copy_(tensor.detach(), non_blocking=True)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/batched_dynamicemb_function.py:57]: Expected

File: examples/commons/distributed/batch_shuffler.py [hunk 2]
```diff
@@ -19,6 +19,44 @@
...
+    target_device = torch.device(device)
+    cpu = torch.as_tensor(values, dtype=torch.int64, device=torch.device("cpu"))
+    if target_device.type != "cuda":
+        return cpu.to(device=target_device)
+    pinned = torch.empty(
+        tuple(cpu.shape),
+        dtype=cpu.dtype,
+        device=torch.device("cpu"),
+        pin_memory=True,
+    )
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Expected. The consumer stream shuold be aware of it. But currently we implicitly agree that no alternate stream

**Comment by JacoCheung** [comment]
:white_check_mark: **Pipeline #53043685** -- success

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104152) |
| train_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104153) |
| inference_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104154) |
| tritonserver_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104155) |
| build_whl | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104156) |
| dynamicemb_test_fwd_bwd_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104157) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104158) |
| unit_test_1gpu_a100 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104159) |
| unit_test_1gpu_h100 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104160) |
| unit_test_4gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104161) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104162) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104163) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104164) |
| inference_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/329104165) |

**Result: 13/14 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/53043685)

File: examples/commons/distributed/batch_shuffler.py [hunk 3]
```diff
@@ -19,6 +19,44 @@
 _SHUFFLE_WITH_ALL2ALL = os.environ.get("SHUFFLE_WITH_ALL2ALL", "0") == "1"
 
 
+def _copy_cuda_tensor_to_pinned_cpu(tensor: torch.Tensor) -> torch.Tensor:
```
Comment (shijieliu) [inline examples/commons/distributed/batch_shuffler.py]: this is duplicated with dynamicemb `def _copy_cuda_tensor_to_pinned_cpu(tensor: torch.Tensor) -> torch.Tensor:`

File: examples/commons/distributed/batch_shuffler.py [hunk 4]
```diff
@@ -19,6 +19,44 @@
...
+    cpu.copy_(tensor.detach(), non_blocking=True)
+    torch.cuda.current_stream(tensor.device).synchronize()
+    return cpu
+
+
+def _tensor_to_cpu_list(tensor: torch.Tensor) -> Any:
+    return _copy_cuda_tensor_to_pinned_cpu(tensor).tolist()
+
+
+def _int64_tensor_from_cpu_nested_list(
```
Comment (shijieliu) [inline examples/commons/distributed/batch_shuffler.py]: this helper is too specific. cna we make this more general? like `create_pinned_tensor`?

File: examples/commons/perf_model/partitioner.py [hunk 5]
```diff
@@ -38,10 +39,85 @@
     Tensor = None
     nvtx = None
 
+# Optional C++ accelerator. Same output as the Python implementation but
```
Comment (shijieliu) [inline examples/commons/perf_model/partitioner.py:42]: if python implementation is only keep as baseline, maybe we can remove it? I believe the benefits of using cpp has already been included in our ablation test using the Shuffler.

File: examples/commons/distributed/batch_shuffler.py (hunk 3 above)
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Yep. I'm just doing it intentionally here because 
keeping the file only under either dynamicemb or examples/hstu seems inappropriate:
- Dynamicemb is the upstream lib, so it should not import any package from `examples/*`
- This functionality seems not semantically related to `embedding`. (In our use case, we are dealing with shuffler but import  from `dynamicemb` ...)

Same situation with [gpu_zipf](https://github.com/NVIDIA/recsys-examples/blob/3c05d0a16313438a619dfa9509740a93cbb0b363/examples/commons/datasets/hstu_batch.py#L161), but this distribution is somewhat related to embedding, and there is a [fallback](https://github.com/NVIDIA/recsys-examples/blob/3c05d0a16313438a619dfa9509740a93cbb0b363/examples/commons/datasets/hstu_batch.py#L164)

File: examples/commons/distributed/batch_shuffler.py (hunk 4 above)
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: I agree. In that case, would it be more natural to move them into `examples/commons`? 
Or, if we want to fix the duplication above, should we move them to `dynamicemb` for consistency instead?

File: examples/commons/perf_model/partitioner.py (hunk 5 above)
Comment (JacoCheung) [reply examples/commons/perf_model/partitioner.py:42]: Maybe we can remove them in the next release? Currently the C++  correctness check requires a python reference. I'd like to keep them at least for one transition release so we're able to do perf/correctness check. The c++ is always the default path.

File: examples/commons/distributed/batch_shuffler.py (hunk 4 above)
Comment (shijieliu) [reply examples/commons/distributed/batch_shuffler.py]: i think it's okay to move it to examples/commons

**Review by shijieliu — APPROVED** [review]
